airbyte-integrations/connectors/source-adaptive/source_adaptive/adaptive/exportData.py

```diff
@@ -0,0 +1,131 @@
+from jinja2 import Template
+
+from datetime import datetime
+from typing import Generator
+
+from source_adaptive.adaptive.base import Adaptive
+from airbyte_cdk.models import (
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    Type,
+)
+from datetime import datetime
+
+
+class AdaptiveExportData(Adaptive):
+
+    def _generate_mapping_order_for_row(self) -> dict:
+        """
+        Since data is provided as csv and we read each record as list,
+        it is helpful to create a mapping that maps where in the list
+        is the respecitve column.
+        """
+
+        # first three columns are static and they are always retrieved
+        mapping_order = {"account_name": 0, "account_code": 1, "level_name": 2}
+
+        # now create a column for each dimensions that is requested with the order respected
+        n_elems = len(mapping_order)
+        dimensions = self.config["method_obj"]["dimensions"]
+        mapping_order.update({d.replace(" ", "_").lower(): n + n_elems for n, d in enumerate(dimensions)})
+
+        # the final column in row is attributed to the amount
+        n_elems = len(mapping_order)
+        mapping_order.update({"amount": n_elems})
+
+        return mapping_order
+
+    def construct_payload(self):
+        """
+        Generate the xml that is sent to the request using jinja templating
+        """
+
+        TEMPLATE = """<?xml version='1.0' encoding='UTF-8'?>
+        <call method="{{method_obj["method"]}}" callerName="Airbyte - auto">
+            <credentials login="{{username}}" password="{{password}}"/>
+            <version name="{{method_obj["version"]}}" isDefault="false"/>
+            <format useInternalCodes="true" includeCodes="false" includeNames="true" displayNameEnabled="true"/>
+            <filters>
+                <accounts>
+                    {% for acc in method_obj["accounts"] -%}
+                    <account code="{{acc}}" isAssumption="false" includeDescendants="true"/>
+                    {% endfor -%}
+                </accounts>
+                <timeSpan start="{{method_obj["date_selected"]}}" end="{{method_obj["date_selected"]}}"/>
+            </filters>
+            <dimensions>
+                {% for dim in method_obj["dimensions"] -%}
+                <dimension name="{{dim}}"/>
+                {% endfor -%}
+            </dimensions>
+            <rules includeZeroRows="false" includeRollupAccounts="true" timeRollups="false">
+                <currency override="USD"/>
+            </rules>
+        </call>"""
+
+        payload = Template(TEMPLATE).render(**self.config)
+        return payload
+
+    def generate_table_name(self) -> str:
+        return "exportData" + "_" + self.config["method_obj"]["version"]
+
+    def generate_table_schema(self) -> dict:
```

Comment:
Magic 🥇

Reply:
manipulating (editing) dictionaries is a pain, update method of dictionary makes it easier